### DISCUSSION TITLE
Sharing self hosted runners across repositories

### DISCUSSION THREAD

**Discussion by manoaman** [community/community#26318] [Actions]
<p>Hi,</p>
<p>I would like to set up my self hosted runner to work with not only a single repository but multiple repositories.  And then I came across there is a organizational level runner solution.</p>
<aside class="quote" data-post="1" data-topic="16876">
  <div class="title">
    <div class="quote-controls"></div>
    
    <a href="https://github.community/t/self-hosted-runners-specific-to-repository/16876">Self-hosted runners specific to repository?</a> <a class="badge-wrapper  bullet" href="/c/code-to-cloud/github-actions/41"><span class="badge-category-parent-bg" style="background-color: #6a737d;"></span><span class="badge-category-bg" style="background-color: #2188ff;"></span><span style="" data-drop-close="true" class="badge-category clear-badge" title="Discussion and questions about GitHub Actions—automation for all of your development workflows, now with world-class CI/CD.">GitHub Actions</span></a>
  </div>
  <blockquote>
    Can self-hosted runners be shared across organization repositories and do we have to create one per repository? It will also be useful to document how to run config.sh in non-interactive mode or automatic provisioning. Thanks
  </blockquote>
</aside>

<p>Currently, I’m not setting up my repos to be part of any organization.  Do I need to create an organization and transfer my account to use this feature?  Also, is there another way to use a self-hosted runner for multiple repositories?  Do I need to create separate directories, “actions-runner”, “actions-runner2”, and then configure separately and launch services for each repository?</p>
<p>Please let me know what is the best practice here.</p>
<p>Thank you,<br>
manoaman</p>

**Comment by Yanjingzhu** [comment, marked as answer]
<p>To use the organization level self-hosted runner , yes, you could create an organization. And then  <a href="https://help.github.com/en/github/administering-a-repository/transferring-a-repository" rel="nofollow noopener">transfer your current repos</a> in your personal account to that organization.<br>
Or you could turn your personal account into an organization directly, please pay attention to the warning before you turn it.<br>
There is the document for “<a href="https://help.github.com/en/github/setting-up-and-managing-your-github-user-account/converting-a-user-into-an-organization" rel="nofollow noopener">Converting a user into an organization</a>” .</p>
<blockquote>
<p>Also, is there another way to use a self-hosted runner for multiple repositories? Do I need to create separate directories, “actions-runner”, “actions-runner2”, and then configure separately and launch services for each repository?</p>
</blockquote>
<p>There is no other ways to use self-hosted runner for multiple repositories.<br>
If you install multiple runners into one machine, creating separate directories for different runners is a preferred way.  Each runner should be configured separately, the url provided for config.sh is your org url not repo url.</p>
<pre><code class="lang-auto">./config.sh --url https://github.com/{orgname}--token ****
</code></pre>
<p>Then you could manage repository permissions in organization settings &gt; actions &gt; self-hosted runner section.<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/d/d6ba3fc5b3807bcc17b048cbd247a61e0f021a36.png" data-download-href="/uploads/short-url/uDzdmAKZ9BzXake8TJioe3nenrM.png?dl=1" title="org agent" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181096879-a4ddd7f7-cbb2-46fb-8bec-973565a50504.png" alt="org agent" data-base62-sha1="uDzdmAKZ9BzXake8TJioe3nenrM" width="690" height="474" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/d/d6ba3fc5b3807bcc17b048cbd247a61e0f021a36_2_690x474.png, https://aws1.discourse-cdn.com/github/optimized/2X/d/d6ba3fc5b3807bcc17b048cbd247a61e0f021a36_2_1035x711.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/d/d6ba3fc5b3807bcc17b048cbd247a61e0f021a36.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/d/d6ba3fc5b3807bcc17b048cbd247a61e0f021a36_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">org agent</span><span class="informations">1065×732 49.7 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div><br>
After you launch the runner service, all repos which you give permission could access to this runner.

**Comment by manoaman** [comment]
<p>Thank you yajingzhu.  I have managed to set up runners for each repo instead of making my account to organization.  Hopefully there is a way to share a runner across my repos without changing my account to organization.</p>

**Reply by SowmyaPhilips** [reply]
Hi can you please tell how to use single runner for multiple repos in an organization

**Reply by hugo-abdou** [reply]
no one replys that not funy guys please wee nead this feature

**Reply by Antiz96** [reply]
The *somewhat sketchy* but functioning workaround is to create one runner per repo in separate directories.
On my side it looks like this:
```
/"somedir"/repo1/runner1
/"somedir"/repo2/runner2
/"somedir"/repoX/runnerX
[...]
```

It's clearly not optimal to have as many runners as there are repos where you could have only one runner shared accros all repositories but, while waiting for GitHub to implement such feature for "regular account/user", it's the only way that I'm aware of.

It's important to notice that each runner will result in a service running on the machine and will also occupy some disk space (~ 500MB per runners *after deleting the action-runner archive*), so make sure your machine have enough RAM and disk space to handle them.
As an example, 9 runners (for 9 repos) on my machine takes about 5GB of disk space and around 700/800MB of RAM at idle.
Basically, any resources normally needed for a single runner (CPU, RAM, disk space) will be multiplied by your number of repos' runners (obviously).

Another important thing to notice is that, if multiple jobs are started at the same time accros multiple repos, they'll all run simultaneously since each repo has its own runner (despite them being hosted on the same machine). So it's worth noticing that, with this workaround, there's no "queue" management like it would be the case with one runner shared accros multiple repositories.
So keep in mind that triggering multiple jobs accros multiple repos at once will get them to run simultaneously on the machine that hosts your runners, provoking eventual conflict (depending on your jobs' actions) and/or performance issues.

Since I'm currently alone working on my repositories, I do not push nor trigger jobs to multiple repos at the same time so I'm personally not really affected by the above but it's still worth mentioning for people who could be affected by that.

Anyway, as I said, that's a *sketchy* but functioning workaround... Clearly not optimal but it does the job while waiting for GitHub to implement a "share runner" feature accros repositories for regular users. Let's hope it happens at some point :slightly_smiling_face:

**Comment by ukreddy-erwin** [comment]
<p>I there any new feature implemented, so that can share a runner accross multiple repositories instead of creating a organization runner?</p>

**Reply by nebuk89** [reply]
This isn't something we are going to do any time soon sadly, I would love to get to the point where we have runner groups for personal name spaces but this isn't something this calendar year. 
We are going to make a change to allow free teams to have more than one runner group to try and ease some of this though :)

**Reply by matt-clarkson** [reply]
Hi there. It's nearly 12 months on since the above. Is there anything on the roadmap yet for sharing runners between repos?

**Reply by mckayreedmoore** [reply]
Likewise wondering if there is any update

**Comment by venkateshbabud** [comment]
Can we run GitHub actions jobs parallelly in different repos, when we are using org level runner? whenever we push the changes into 2 different repos, 1st job will get trigger and 2nd job will be in queue and wait for 1st job to complete. but can we run both the jobs at same time?

**Reply by maxkratz** [reply]
AFAIK, one (1) GitHub Actions runner can only run one (1) job at a time, regardless of the scenario (different repositories, org, ...).

**Reply by nebuk89** [reply]
Yeah today the runner only runs one job at a time :)

**Comment by caseyjkey** [comment]
Any updates on this? It's not resource-efficient to have one runner per repo.

**Comment by sdawka** [comment]
For anyone who needs it, I made a tool for it after reading some suggestions from this thread. Thanks for all the discussion - https://github.com/sdawka/gh-runnermaxxer
